## Ομάδα 

Στοιχεία Μέλους 1: Στεφανάκης Γεώργιος (el18436)

Στοιχεία Μέλους 2: Μπούφαλης Οδυσσεύς - Δημήτριος (el18118)

# Τεχνητή Νοημοσύνη: Εργαστηριακή Άσκηση 2
---

Ο στόχος της εργασίας είναι η κατασκευή ενός συστήματος προτάσεων (Recommendation System) για ταινίες. Οι προτάσεις αυτές θα πηγάζουν τόσο από τα χαρακτηριστικά της ταινίας όσο και από ορισμένες αξιολογήσεις του κάθε χρήστη.

Στα δεδομένα της άσκησης περιλαμβάνονται ένα αρχείο με το όνομα movies_metadata.csv το όποιο περιέχει τα χαρακτηριστικά κάθε ταινίας όπως θέμα, σκηνοθέτης ηθοποιοί, λέξεις κλειδιά κ.α. από το imdb καθώς και τα αρχεία ratings.csv τα όποια περιέχουν πραγματικές αξιολογήσεις χρηστών, χωρισμένες σε train και σε test.


# Εκφώνηση
Στην παρούσα εργασία σας ζητείται να μελετήσετε και να υλοποιήσετε τα παρακάτω:<br>

### Μέρος 1
Στο μέρος 1 και 2 θα εργαστείτε μόνο με το αρχείο movies_metadata.csv, ενώ στο μέρος τρία θα δουλέψετε και με τα αρχεία των αξιολογήσεων.
#### Ερώτημα 1α

Αρχικά θα πρέπει, αφού μελετήσετε τη δομή και τα χαρακτηριστικά του movies_metadata.csv, να κατασκευάσετε μια βάση γνώσης για την Prolog η όποια ουσιαστικά θα αποτελεί τον κόσμο με τον όποιο θα εργαστείτε στην συνέχεια. Τα κατηγορήματα που θα δημιουργηθούν θα σας βοηθήσουν και στην κατασκευή του recommender και θα είναι της μορφής:

```
director(Movie, Director).
genre(Movie, Genre).
```
#### Ερώτημα 1β

Αφού δημιουργήσετε τον κόσμο του προβλήματος, στη συνέχεια καλείστε να δημιουργήσετε, σε Prolog, απλούς κανόνες οι οποίοι θα βρίσκουν όλες τις ταινίες με:
1.	Κοινό θέμα (κάποιες λέξεις σχετικά με το genre κοινές)
2.	Αρκετά κοινό θέμα (κάποιες λιγότερες λέξεις σχετικά με το genre κοινές π.χ. 3)
3.	Σχετικά κοινό θέμα (λίγες λέξεις σχετικές με το genre κοινές π.χ. 1)
4.	Κοινός σκηνοθέτης
5.	Ακριβώς ίδια πλοκή (κάποιες λέξεις κλειδιά της πλοκής κοινές)
6.	Σχετικά ίδια πλοκή (κάποιες λιγότερες λέξεις κλειδιά κοινές)
7.	Ίδιους τους βασικούς ηθοποιούς (και τους 3)
8.	Αρκετά ίδιους βασικούς ηθοποιούς (ορισμένους βασικούς ηθοποιούς κοινούς π.χ. 2)
9.	Σχετικά ίδιους ηθοποιούς (π.χ. 1 από τους 3)
10.	Ίδια γλώσσα
11.	Είναι έγχρωμες ή ασπρόμαυρες
12. Κοινό studio παραγωγής
13. Κοινή χώρα παραγωγής 
14. Ίδια δεκαετία 

Αξίζει να σημειωθεί ότι στα παραπάνω μπορείτε να προσθέσετε περισσότερα ερωτήματα ή να αλλάξετε την κλιμακωσιμότητα των queries (πέρα από το ίδιο, αρκετά ίδιο, σχετικά ίδιο) που θα κατασκευάσετε(π.χ. μια κλίμακα από 1 στα 5 όπου αυτό είναι δυνατόν), μιας και αυτά στην συνέχεια θα χρησιμοποιηθούν για την λειτουργία του recommender. Έτσι μπορείτε να προσθέσετε queries με τα όποια μπορεί να παράγονται καλύτερες συστάσεις  (το αρχείο movie_metadata.csv περιέχει πολλές πληροφορίες ακόμα για κάθε ταινία όπως έτος κυκλοφορίας, βαθμολογία στο imdb, facebοok_likes κ.α.). Το συγκεκριμένο μέρος εργασίας είναι προπαρασκευαστικό οπότε όσο καλύτερα και πλουσιότερα τα ερωτήματα που θα φτιάξετε σε αυτό το μέρος τόσο καλύτερη θα είναι η απόδοσή των συστημάτων συστάσεων των επόμενων ερωτημάτων.

Περισσότερες πληροφορίες για το dataset μπορείτε να διαβάσετε σε αυτό το [Link](https://www.kaggle.com/georgefila/movies-metadata):
### Μέρος 2: Recommendation System
Στο σημείο αυτό καλείστε με βάση αυτά που κάνατε στο μέρος 1 να κατασκευάσετε queries τα όποια θα σας επιστρέφουν παρόμοιες (σε χαρακτηριστικά) ταινίες. Τα ερωτήματα αυτά θα είναι κλιμακούμενα, δηλαδή θα υπάρχουν ερωτήματα που επιστρέφουν αρκετά κοινές ταινίες αλλά και που επιστρέφουν λιγότερο και λιγότερο κοινές (σε μια κλίμακα π.χ. από 1 σε 5). Για παράδειγμα:

```
find_simmilar_movies_5("Pirates Of The Caribbean", M).
M = "Pirates Of The Caribbean: On Stranger Tides"
M = "The Chronicles Of Narnia"
M = "Prince Of Persia: The Sands Of Time"
...
```
Για παράδειγμα, το παραπάνω ερώτημα θα επιστρέφει αρκετά κοινές σε περιεχόμενο ταινίες με την ταινια "Pirates Of The Caribbean". Θα υπάρχουν και αντίστοιχα ερωτήματα που θα βρίσκουν λιγότερο όμοιες ταινίες. Ο δείκτης ομοιότητας των ταινιών είναι αυθαίρετος και μπορείτε να τον ορίσετε εσείς όπως θέλετε, αρκεί να υπάρχει κάποια λογική σύνδεση με τα δεδομένα που περιέχονται στο αρχείο movies_metadata.csv. 

Συνεπώς η συνάρτηση που θα κάνει τις προτάσεις (recommendation) με είσοδο μια ταινία πρέπει να επιστρέφει-εκτυπώνει μια λίστα με τις προτεινόμενες ταινίες κατά φθίνουσα σειρά ομοιότητας. 


###3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις - Αξιολογήσεις του χρήστη


Σε αυτό το σημείο θα εργαστείτε με τα αρχεία ratings τα όποια περιέχουν αξιολογήσεις (από 1 μέχρι 5) για τις παραπάνω ταινίες. Το προηγούμενο σύστημα συστάσεων προτείνει στον χρήστη ταινίες αποκλειστικά με βάση την ομοιότητά τους. Σε αυτό το σημείο θα γίνει μια αναβάθμιση του συστήματος έτσι ώστε να παράγονται καλύτερες  συστάσεις οι όποιες θα λαμβάνουν υπόψιν και τις προτιμήσεις του χρήστη, οι όποιες θα εξάγονται από τις αξιολογήσεις που έχει κάνει μέχρι στιγμής. 

Η εκπαίδευση του recommender θα γίνεται ως εξής:

Για κάθε ταινία θα υπάρχει ένα score το όποιο αρχικά θε είναι ίσο με 0 και θα διαμορφώνεται από τις αξιολογήσεις κάθε user. Έτσι για έναν χρήστη με βάση τις αξιολογήσεις που υπάρχουν στο αρχείο train_ratings θα πρέπει:

1.	Για κάθε ταινία που έχει βαθμολογήσει να βρίσκονται οι κοινές ταινίες ανά κλίμακα και στο μέχρι τώρα σκορ κάθε παρόμοιας ταινίας θα προστίθεται ένα βάρος το όποιο θα μπορούσε να είναι το ποσοστό ομοιότητας της ταινίας (δηλαδή ένα βάρος για κάθε κλίμακα, αν δύο ταινίες μοιάζουν ενισχύουμε το βάρος που προσθέτουμε από το να μοιάζουν λιγότερο) επί τον βαθμό που έχει βάλει ο χρήστης για την αρχική ταινία (διαφορετικό είναι ο χρήστης να έχει βάλει 5/5 ή 1/5 σε μια ταινία από 3/5). 

2. Στη συνέχεια, ανάλογα με το σκορ που έχει σχηματιστεί για κάθε ταινία, θα επιλέγεται αν αυτή θα μπορούσε να είναι προτεινόμενη για τον χρήστη ή όχι και θα μετράμε το πόσο καλά τα πήγε το σύστήμά μας με βάση ορισμένες μετρικές.

Η λογική πίσω από την παραπάνω διαδικασία είναι ότι παρόμοιες ταινίες θα έχουν ανάλογο βαθμό. Για παράδειγμα αν ένας χρήστης έχει αξιολογήσει αρκετές  ταινίες οι όποιες είναι sci-fiction με 5/5 τότε μια ταίνια sci-fiction την όποια δεν έχει δεί λογικά θα του αρέσει και θα έπρεπε να την προτείνουμε.

Μετά την εκπαίδευση του συστήματος σας καλείστε να δοκιμάσετε τον recommender που κατασκευάσατε στην πράξη. Για τον σκοπό αυτό θα φορτώσετε το αρχείο test_ratings.csv όπου περιέχονται οι αξιολογήσεις του ίδιου χρήστη για άλλες ταινίες. Το σύστημά σας πρέπει να προβλέπει αν μια ταινία θα πρέπει να προταθεί στον χρήστη. Μια ταίνια θα έπρεπε να έχει προταθεί στον  χρήστη, αν έχει βαθμό μεγαλύτερο του 3. Συνεπώς για την επίβλεψη του συστήματός σας θα πρέπει για κάθε μια από τις ταινίες του αρχείου test_ratings.csv να επιστρέφεται 1 ή 0 αν η ταινία θα έπαιρνε βαθμό μεγαλύτερο του 3 ή όχι, δηλαδή αν θα έπρεπε να την είχαμε προτείνει ή όχι.

Στην συνέχεια, σε συνδιασμό με τις πραγματικές απαντήσεις του χρήστη θα αξιολογήσετε το σύστημά σας χρησιμοποιόντας τις μετρικές: precision, recall, f1 οι οποίες είναι οι πλέον γνωστές μετρικές και ευρέως χρησιμοποιούμενες τεχνικές για την επίβλεψη-μέτρηση απόδοσης ανάλογων συστημάτων.

1. Precision: Δείχνει πόσο ακριβές είναι το σύστημα. Υπολογίζει πόσα από τα στιγμιότυπα τα όποια προβλέψαμε ότι ανήκουν σε μια κλάση όντως ανήκουν σε αυτή. Η μετρική αυτή μας δίνει μια εικόνα σχετικά με τον αριθμό των ταινίων που προβλέψαμε ως προτεινόμενες ενώ δεν θα έπρεπε.

2. Recall: Υπολογίζει πόσα από τα στιγμιότυπα που ανήκουν σε μια κλάση (π.χ. προτεινόμενες ταινίες) προβλέφθηκαν σωστά.

3. F1: Είναι ένας μέσος μεταξύ των παραπάνω δυο μετρικών, έτσι ώστε να διατηρείται μια ισορροπία μεταξύ τους. Υπολογίζεται από την παρακάτω σχέση:

$$F_1=2\frac{Precision\times{Recall}}{Precision+Recall}$$

Οι παραπάνω συναρτήσεις παρέχονται από την βιβλιοθήκη scikit-learn.

Τέλος για την καλύτερη επίβλεψη του συστήματος σας μπορείτε να εκπαιδεύσετε  τον recommender σας με ένα υποσύνολο ταινίων από ελάχιστες, λίγες μέχρι και πολλές (π.χ. 3, 5, 10, 50, ...) για να μελετήσετε κατά πόσο σας βοηθούν οι επιπλέον αξιολογήσεις κάθε φορά (δηλαδή κατά πόσο βελτιώνονται οι παραπάνω μετρικές στο test set). 

Έτσι π.χ. μπορείτε να εντοπίσετε περιπτώσεις όπως για παράδειγμα ότι με έναν recommender μπορεί να μην επιτυγχάνετε πολύ υψηλό σκορ όσο με άλλους, αλλά το βέλτιστο σκόρ σας επιτυγχάνεται πολύ γρηγόρα π.χ. με μόνο 10 ταινίες αντί 100. Έτσι για παράδειγμα αν ο αλγόριθμος τα πηγαίνει πολύ καλά για τρεις ταινίες και στη συνέχεια το σκορ βελτιώνεται ελάχιστα τότε αυτός ενδείκνυται για ένα σύστημα συστάσεων για νέους χρήστες όπου δεδομένου λίγων ταινιών ο αλγόριθμος είναι σε θέση να το να προτείνει καλές συστάσεις. Ενώ αν η καλύτερη απόδοση του αλγορίθμου σας είναι βέλτιστη με περισσότερες ταινίες π.χ. 50 τότε αυτός ο αλγόριθμος συστάσεων ενδείκνυται για παλιούς χρήστες με πολλές αξιολόγησεις. 

---
## Κατασκευή Περιβάλλοντος Εργασίας
---
### Διάβασμα Αρχείων στο Colab (Μόνο για το Colab)
Αν η υλοποίηση γίνει στο google colab τότε μπορεί να χρησιμοποιηθεί το google drive ως file system. Για να γίνει Mount το google drive τρέχουμε τον παρακάτω κώδικα και κλικάρουμε στο link που θα μας εμφανιστεί. 

```
from google.colab import drive
drive.mount('/content/drive')
import os
os.listdir('/content/drive/My Drive')
```
Έπειτα στην σελίδα που άνοιξε επιλέγουμε το mail μας και στο επόμενο παράθυρο που θα μας ανοίξει πατάμε Να επιτρέπεται. Στην συνέχεια αντιγράφουμε τον κωδικό που θα μας βγάλει και τον κάνουμε paste στο Input που έχει ανοίξει στο colab. Έτσι πλέον αν έχουμε ανεβάσει ένα αρχείο στο google drive μπορούμε να το βρούμε στην θέση:

```
movies_filename = '/content/drive/My Drive/' + movies_metadata.csv
```
Μπορούμε πλέον κανονικά να δουλέψουμε φτιάχνοντας φακέλους ή αρχεία και γενικότερα κάνοντας οτιδήποτε θα κάναμε αν ήμασταν τοπικά. 

### Prolog μέσω Python

Το πακέτο που θα χρησιμοποιηθεί για την επικοινωνία Python και Prolog είναι το pyswip (https://pypi.org/project/pyswip/). Για να δουλέψει το Pyswip  χρειάζεται να υπάρχει το Swi-Prolog το όποιο αν δουλεύουμε τοπικά πρέπει να το εγκαταστήσουμε, ακολουθώντας αντίστοιχες οδηγίες στην σελίδα του εργαλείου. Για να γίνει του Swi-Prolog η εγκατάσταση στο Google Colab πρέπει να τρέξουμε τον παρακάτω κώδικα:

```
!sudo apt-get install software-properties-common
!sudo apt-add-repository ppa:swi-prolog/stable
!sudo apt-get update
!sudo apt-get install swi-prolog
```
Σε κάποιο σημείο της εκτέλεσης εμφανίζεται ένα μήνυμα ότι πρέπει να πατήσουμε enter σε ένα input για να συνεχίσει η διαδικασία. Έπειτα από αυτό η εκτέλεση θα συνεχίσει χωρίς κάποιο πρόβλημα.

Τέλος πρέπει να εγκαταστήσουμε το pyswip (**όπου και να δουλεύουμε**) όπως παρακάτω:

```
!pip isntall pyswip
```


---
#Κώδικας για κατασκευή περιβάλλοντος εργασίας

##Μόνο για Google Colab

Κώδικας για να γίνει Mount to Google Drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive/', force_remount=True)

import os
path="./drive/MyDrive/ai-lab-2/" # με ευκόλο τρόπο μπορείτε να αλλάξετε το path που είναι αποθηκευμένα τα αρχεία σας
os.listdir(path)

Mounted at /content/drive/


['.ipynb_checkpoints',
 'db.pl',
 'train_ratings.csv',
 'test_ratings.csv',
 'movies_metadata.csv',
 '1b.pl',
 '2.pl']

In [ ]:
%%capture
#install swi-prolog
!sudo apt-get install software-properties-common
!sudo apt-add-repository ppa:swi-prolog
!sudo apt-get update
!sudo apt-get install swi-prolog
#install pyswip
!pip install pyswip


# **Μέρος 1: Μελέτη των Metadata, Δημιουργία κόσμου και των βασικών queries.**


In [ ]:
import pandas as pd
from pyswip import Prolog
import json
import math

In [ ]:
# Η βιβλιοθήκη pandas είναι χρήσιμη για την εργασία με τέτοια δεδομένα
import pandas as pd 
# Διάβασμα του αρχείου 'movie_metadata.csv' 
data = pd.read_csv(path + "movies_metadata.csv") 
#Στο csv υπαρχούν κελία με nan τιμές
#Στις θέσεις αυτές βάζουμε 'UNK' πράγμα που  κάνουμε με την παρακάτω συνάρτηση
data.fillna("UNK", inplace=True)
# Preview the first 5 lines of the loaded data 
data.head()

Unnamed: 0     budget  ...      actor_2_name       actor_3_name
0           0  237000000  ...  Sigourney Weaver       Stephen Lang
1           1  300000000  ...   Keira Knightley  Stellan Skarsgård
2           2  245000000  ...       Léa Seydoux      Ralph Fiennes
3           3  250000000  ...       Gary Oldman      Anne Hathaway
4           4  260000000  ...   Samantha Morton       Willem Dafoe

[5 rows x 27 columns]

In [ ]:
def clean_text(text):
  text = text.replace(u'\xa0', u'')
  text = text.replace(u"'", u'')
  return text

In [ ]:
columns = {
    'budget': 'budget',
    'genres': 'genre',
    'plot_keywords': 'keywords',
    'language': 'language',
    'original_title': 'originalTitle',
    'popularity': 'popularity',
    'production_companies': 'productionCompany',
    'production_countries': 'productionCountry',
    'release_date':'releaseMonth',
    'gross': 'gross',
    'duration': 'duration',
    'spoken_languages': 'spokenLanguages',
    'vote_average': 'voteAverage',
    'title_year': 'titleYear',
    'country': 'country',
    'director_name': 'director',
    'actor_1_name': 'actor1' ,
    'actor_2_name': 'actor2',
    'actor_3_name': 'actor3'
}

def parse_single_column(col, file):
  literals = []

  for row in data.itertuples(index = True, name = 'Pandas'):
    movie_title = clean_text(getattr(row, 'movie_title'))
    attribute = clean_text(str(getattr(row, col)))
    if attribute == 'UNK': continue

    if col == 'release_date' and len(attribute.split('-')) > 1:
      month = attribute.split('-')[1]
      literals.append(columns["release_date"] + "('" + movie_title + "','" + month +"')")
    else:
      for val in attribute.split('|'):
        if col == 'title_year':
          val = float(val)
          math.floor(val)
          val = int(val)
          val = str(val)
        literals.append(columns[col] + "('" + movie_title + "', '" + val + "')")

  literals.sort()

  for l in literals:
    file.write(l + ".\n")

def parse_json_column(col, file):
  literals = []

  for row in data.itertuples(index = True, name = 'Pandas'):
    movie_title = clean_text(getattr(row, 'movie_title'))
    attributes = getattr(row, col)
    if '"' in attributes: continue
    if attributes == 'UNK': continue
    attributes = json.loads(clean_text(attributes.replace("'", '"')))

    for atr in attributes:
      atr = clean_text(atr['name'])
      literals.append(columns[col] + "('" + movie_title + "','" + atr + "')")

  literals.sort()
  for l in literals:
    file.write(l + ".\n")

file = open(path + 'db.pl', 'w', encoding='utf-8')
for col in columns:
  if col not in ["production_companies", "production_countries","spoken_languages"]:
    parse_single_column(col, file)
  else:
    parse_json_column(col, file)

file.close()


In [ ]:
# prolog.assertz('(directed_by(X,Y) :- findall(M,director(M,X),Y))')

In [ ]:
f = open(path + "1b.pl", 'w')
f.write(
  '''
  fiveCommonGenres(X, Y) :-
    genre(X, N1),
    genre(X, N2), N2 \= N1,
    genre(X, N3), N3 \= N1, N3 \= N2,
    genre(X, N4), N4 \= N1, N4 \= N2, N4 \= N3,
    genre(X, N5), N5 \= N1, N5 \= N2, N5 \= N3, N5 \= N4,
    genre(Y, N1),
    genre(Y, N2),
    genre(Y, N3),
    genre(Y, N4),
    genre(Y, N5),
    X \= Y.

  fourCommonGenres(X, Y) :-
    genre(X, N1),
    genre(X, N2), N2 \= N1,
    genre(X, N3), N3 \= N1, N3 \= N2,
    genre(X, N4), N4 \= N1, N4 \= N2, N4 \= N3,
    genre(Y, N1),
    genre(Y, N2),
    genre(Y, N3),
    genre(Y, N4),
    X \= Y.

  threeCommonGenres(X, Y) :-
    genre(X, N1),
    genre(X, N2), N2 \= N1,
    genre(X, N3), N3 \= N1, N3 \= N2,
    genre(Y, N1),
    genre(Y, N2),
    genre(Y, N3),
    X \= Y.

  twoCommonGenres(X, Y) :-
    genre(X, N1),
    genre(Y, N1),
    genre(X, N2), N2 \= N1,
    genre(Y, N2),
    X \= Y.

  oneCommonGenre(X, Y) :-
    genre(X, N1),
    genre(Y, N1),
    X \= Y.

  commonDirector(X, Y):-
    director(X, D),
    director(Y, D),
    X \= Y.

  fiveCommonKeywords(X, Y) :-
    keywords(X, N1),
    keywords(X, N2), N2 \= N1,
    keywords(X, N3), N3 \= N1, N3 \= N2,
    keywords(X, N4), N4 \= N1, N4 \= N2, N4 \= N3,
    keywords(X, N5), N5 \= N1, N5 \= N2, N5 \= N3, N5 \= N4,
    keywords(Y, N1),
    keywords(Y, N2),
    keywords(Y, N3),
    keywords(Y, N4),
    keywords(Y, N5),
    X \= Y.

  fourCommonKeywords(X, Y) :-
    keywords(X, N1),
    keywords(X, N2), N2 \= N1,
    keywords(X, N3), N3 \= N1, N3 \= N2,
    keywords(X, N4), N4 \= N1, N4 \= N2, N4 \= N3,
    keywords(Y, N1),
    keywords(Y, N2),
    keywords(Y, N3),
    keywords(Y, N4),
    X \= Y.

  threeCommonKeywords(X, Y) :-
    keywords(X, N1),
    keywords(X, N2), N2 \= N1,
    keywords(X, N3), N3 \= N1, N3 \= N2,
    keywords(Y, N1),
    keywords(Y, N2),
    keywords(Y, N3),
    X \= Y.

  twoCommonKeywords(X, Y) :-
    keywords(X, N1),
    keywords(Y, N1),
    keywords(X, N2), N2 \= N1,
    keywords(Y, N2),
    X \= Y.

  oneCommonKeywords(X, Y) :-
    keywords(X, N1),
    keywords(Y, N1),
    X \= Y.

  actor(X, N) :- actor1(X, N).
  actor(X, N) :- actor2(X, N).
  actor(X, N) :- actor3(X, N).

  sameActors(X, Y) :-
    actor(X, N1),
    actor(X, N2), N2 \= N1,
    actor(X, N3), N3 \= N1, N3 \= N2,
    actor(Y, N1),
    actor(Y, N2),
    actor(Y, N3),
    X \= Y.

  twoCommonActors(X, Y) :-
    actor(X, N1),
    actor(X, N2), N2 \= N1,
    actor(Y, N1),
    actor(Y, N2),
    X \= Y.

  oneCommonActor(X, Y) :-
    actor(X, N),
    actor(Y, N),
    X \= Y.

  sameLanguage(X, Y) :-
    language(X, N),
    language(Y, N),
    X \= Y.

  blackAndWhite(X) :-
    keywords(X, 'black and white').

  commonCompany(X, Y) :-
    productionCompany(X, N),
    productionCompany(Y, N),
    X \= Y.

  commonCountry(X, Y) :-
    productionCountry(X, N),
    productionCountry(Y, N),
    X \= Y.

  equals(X, X).

  sameDecade(X, Y) :-
    titleYear(X, A),
    titleYear(Y, B),
    atom_number(A, N1),
    atom_number(B, N2),
    P is div(N1, 1000),
    Q is div(N2, 1000),
    equals(P, Q),
    R is div(N1, 10),
    S is mod(R, 10),
    T is div(N2, 10),
    W is mod(T, 10),
    equals(S, W),
    X \= Y.

  highlyVoted(X) :-
    voteAverage(X, A),
    atom_number(A, N),
    N > 7.

  simillarHighlyVoted(X, Y) :-
    twoCommonKeywords(X, Y),
    highlyVoted(Y),
    X \= Y.
  '''
)
f.close()

#### Demo

In [ ]:
prolog = Prolog()
prolog.consult(path + 'db.pl')
prolog.consult(path + '1b.pl')

movies = [
  '2001: A Space Odyssey',
  '2046',
  'Pierrot le Fou',
  'Spellbound',
  'Trainspotting'
]

queries = [
  'twoCommonGenres',
  'twoCommonKeywords',
  'commonDirector',
  'twoCommonActors',
  'sameDecade',
  'commonCountry',
  'simillarHighlyVoted'
]

print('------------')
for m in movies:
  print('Movie: ' + m)
  for q in queries:
    temp = prolog.query(q + "('" + m + "', X)")

    res = [r['X'] for r in temp]
    print('Query: ' + q)
    s = set()
    for r in res:
      s.add(r)
    print('Found ' + str(len(s)) + ' movies. Ten of them are:')
    print(str(list(s)[:10]))
    print('')
  print('------------')


------------
Movie: 2001: A Space Odyssey
Query: twoCommonGenres
Found 264 movies. Ten of them are:
['G.I. Joe: The Rise of Cobra', 'Man of Steel', 'Planet of the Apes', 'Pandorum', 'Race to Witch Mountain', 'Hellboy II: The Golden Army', 'Atlas Shrugged Part II', 'Star Trek III: The Search for Spock', 'Paycheck', 'X-Men: Days of Future Past']

Query: twoCommonKeywords
Found 25 movies. Ten of them are:
['In the Shadow of the Moon', 'The Matrix Reloaded', 'I, Robot', 'The Astronauts Wife', 'Ex Machina', 'Blade Runner', 'Moon', 'Eagle Eye', 'Apollo 18', 'Spaceballs']

Query: commonDirector
Found 5 movies. Ten of them are:
['Eyes Wide Shut', 'Dr. Strangelove or: How I Learned to Stop Worrying and Love the Bomb', 'The Shining', 'Barry Lyndon', 'Lolita']

Query: twoCommonActors
Found 0 movies. Ten of them are:
[]

Query: sameDecade
Found 70 movies. Ten of them are:
['55 Days at Peking', 'Night of the Living Dead', 'Once Upon a Time in the West', 'You Only Live Twice', 'Hud', 'The Great Esca

# **2ο Μέρος: Recommendation System με βάση μόνο τα χαρακτηριστικά των ταινιών.**

Στο σημείο αυτό με βάση τους κανόνες που κατασκευάστηκαν στο Μέρος 1 θα κατασκευαστούν κατηγορήματα για την ομοιότητα ταινιών. Παρακάτω δίνεται ένα μικρό παράδειγμα ενός κανόνα και πώς αυτός θα μπορούσε να γραφτεί μέσω του Pyswip. Επίσης όπως αναφέρεται και σε σχόλιο παραπάνω θα μπορεί να γραφτεί και μια βάση δεδομένων με τους κανόνες και να γίνει απευθείας consult.

Στο παρακάτω παράδειγμα το 5 και το 4 εκφράζουν την ομοιότητα των ταινιών π.χ. οι ταινίες που παράγονται μέσω του find_similar_5 είναι πιο όμοιες από αυτές που παράγονται μέσω του find_similar_4. 


In [ ]:
f = open(path + "2.pl", 'w')
f.write(
  '''
  findSim1(X, Y) :- oneCommonKeywords(X, Y).
  findSim1(X, Y) :- oneCommonGenre(X, Y).
  findSim1(X, Y) :- commonDirector(X, Y).
  findSim1(X, Y) :- commonCountry(X, Y).
  findSim1(X, Y) :- sameLanguage(X, Y).
  findSim1(X, Y) :- sameDecade(X, Y).

  findSim2(X, Y) :- twoCommonKeywords(X, Y).
  findSim2(X, Y) :- twoCommonGenres(X, Y).
  findSim2(X, Y) :- commonDirector(X, Y).
  findSim2(X, Y) :- sameDecade(X, Y), commonCountry(X, Y).
  findSim1(X, Y) :- sameLanguage(X, Y).

  
  findSim3(X, Y) :- threeCommonKeywords(X, Y).
  findSim3(X, Y) :- threeCommonGenres(X, Y).
  findSim3(X, Y) :- sameDecade(X, Y), commonDirector(X, Y).
  findSim3(X, Y) :- oneCommonActor(X, Y).

  findSim4(X, Y) :- twoCommonKeywords(X, Y), twoCommonGenres(X, Y).
  findSim4(X, Y) :- fourCommonKeywords(X, Y).
  findSim4(X, Y) :- sameDecade(X, Y), commonDirector(X, Y).
  findSim4(X, Y) :- twoCommonActors(X, Y).

  findSim5(X, Y) :- threeCommonKeywords(X, Y), threeCommonGenres(X, Y).
  findSim5(X, Y) :- threeCommonKeywords(X, Y), twoCommonGenres(X, Y).
  findSim5(X, Y) :- threeCommonGenres(X,Y),oneCommonActor(X,Y).
  findSim5(X, Y) :- fiveCommonKeywords(X, Y).
  findSim5(X, Y) :- sameDecade(X, Y), commonDirector(X, Y), sameLanguage(X, Y).
  
  '''
)
f.close()

In [ ]:
prolog = Prolog()
prolog.consult(path + 'db.pl')
prolog.consult(path + '1b.pl')
prolog.consult(path + '2.pl')

In [ ]:
movies = [
  '2001: A Space Odyssey',
  '2046',
  'Spellbound',
  'Trainspotting',
  'Avatar'
]

def simple_recommender(movie, precision = 5):
    s = set()
    q = prolog.query("findSim" + str(precision) + "('" + movie + "', M)")
    for soln in q:
        m = soln['M'] 
        if m not in s:
            s.add(soln['M'])
    q.close()
    answers = s
    return answers

print('------------')
for m in movies:
  print('Movie: ' + m)
  for i in range(5, 0, -1):
    print('')
    print('Precision: ' + str(i))
    print(list(simple_recommender(m, i))[:10])
  print('------------')


------------
Movie: 2001: A Space Odyssey

Precision: 5
['Dr. Strangelove or: How I Learned to Stop Worrying and Love the Bomb', 'The Matrix Revolutions', 'Interstellar', 'Transcendence', 'Lolita']

Precision: 4
['Solaris', 'Silent Running', 'Dr. Strangelove or: How I Learned to Stop Worrying and Love the Bomb', 'Armageddon', 'The Matrix Reloaded', 'The Matrix Revolutions', 'Moonraker', 'Interstellar', 'Transcendence', 'Lolita']

Precision: 3
['Solaris', 'Knowing', 'Tomorrowland', 'Inception', 'Star Trek: The Motion Picture', 'Dr. Strangelove or: How I Learned to Stop Worrying and Love the Bomb', 'Beneath the Planet of the Apes', 'Paycheck', 'The Matrix Revolutions', 'Congo']

Precision: 2
['Planet of the Apes', 'Once Upon a Time in the West', 'Hud', 'Mystery Men', 'Paint Your Wagon', 'The Martian', 'Star Wars: Episode I - The Phantom Menace', 'Terminator Genisys', 'Jumper', 'Hot Tub Time Machine']

Precision: 1
['Transporter 2', 'The Big Short', 'The Bourne Identity', 'Battle for the 

In [ ]:
list(simple_recommender('Avatar'))[:5]

['John Carter',
 'Aliens',
 'Pirates of the Caribbean: On Stranger Tides',
 'Predators',
 'Guardians of the Galaxy']

# **3ο Μέρος: Recommendation System Με βάση τις προτιμήσεις-Αξιολογήσεις του χρήστη-Εκπαίδευση και Πρόβλεψη**


Αρχικά μελετάμε τις αξιολογήσεις κάθε χρήστη για να καταλάβουμε την δομή και τις πληροφορίες κάθε αρχείου.




In [ ]:
from tqdm.notebook import tqdm
from sklearn.metrics import precision_score, recall_score, f1_score
import numpy as np
import random

rating_weights = {0: -1, 1: -0.5, 2:0, 3:0, 4:0.5, 5:1}
score_weights = {1:0.05, 2:0.5, 3:3, 4:4, 5:5} # ανάλογα με τα επίπεδα ομοιότητας που έχουν οριστεί στην simple_recommender

def train_recommender(ratings, rating_weights, score_weights, number_of_movies = 10):
    """
    Στην συνάρτηση αυτή μπορούμε να ορίζουμε ποιο υποσύνολο των αξιολογήσεων θα χρησιμοποιήσουμε για το train μαζί με τα βάρη ομοιότητας και σκορ
    Σε συνδυασμό με τον αριθμό των ταινιών που θέλουμε να χρησιμοποιήσουμε σαν σύνολο δεδομένων π.χ. 10 από τις 100 ή 3 από τις 100 κ.ο.κ
    Αν θέλουμε να χρησιμοποιήσουμε όλες τις ταινίες σαν training set τότε ορίζουμε το number_of_movies = - 1
    """

    if number_of_movies > len(ratings):
        number_of_movies = len(ratings)


    if number_of_movies != -1:
        indexes = random.sample(range(len(ratings)), number_of_movies)
        ratings = ratings.iloc[indexes]

    movie_score = {}
    seen_movies = set()
    for row in tqdm(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        similar_movies = {i: simple_recommender(movie, i) for i in range(5, 0, -1)}

        for p in range(5, 0, -1):
          for similar_movie in similar_movies[p]:
            if similar_movie in seen_movies: continue
            else: seen_movies.add(similar_movie)
            if similar_movie not in movie_score:
                movie_score[similar_movie] = rating_weights[int(rating)] * score_weights[p]
            else:
                movie_score[similar_movie] += rating_weights[int(rating)] * score_weights[p]
                #το weight θα το ορίσετε ανα επίπεδο ομοιότητας οι πολύ όμοιες ταινίες θα έχουν μεγαλύτερο βάρος

    return movie_score


#αυτό είναι ένα παράδειγμα για το πως θα μπορούσε να υλοποιήθει η predict
#έχουμε ορίσει ότι μια ταινία θα έπρεπε να είναι προτεινόμενη αν είχε σκορ > 0
def predict_example(ratings, movie_score):
    real, pred = [], []
    for i, row in enumerate(ratings.itertuples(index=True, name='Pandas')):
        movie = clean_text(getattr(row, 'movie_title'))
        rating = getattr(row, 'rating')

        if movie in movie_score: #αν έχουμε σχηματίσει βαθμολογία για την ταινία αυτή
            pred.append(int(movie_score[movie] > -0.5)) #heuristic για το αν μια ταινία είναι προτεινόμενη
            real.append(int(rating > 3))# έτσι ορίζουμε ότι μια ταινία θα έπρεπε να είναι προτεινόμενη
            #η συνθήκη αυτή δεν μπορεί να αλλάξει
        else: #δεν μπορούμε να προτείνουμε κάτι για το όποιο δεν έχουμε σχηματίσει εικόνα
            pred.append(0)
            real.append(int(rating > 3))

    return real, pred


def get_metrics(real, pred):
    metrics = {}
    metrics["precision"] = precision_score(real, pred)
    metrics["recall"] = recall_score(real, pred)
    metrics["f1"] = f1_score(real, pred)
    return metrics



Η παραπάνω εκπαιδεύουν, τεστάρουν και μετρούν την απόδοσή του συστήματος στάσεων μας. Για την εκπαίδευση του συστήματος μπορούμε κάθε φορά να χρησιμοποιήσουμε ένα τυχαίο υποσύνολο του training set. Όμως είναι πιθανό το υποσύνολο των ταινιών αυτόν να επηρεάζει τα αποτελέσματα στο training set. Για παράδειγμα από 3 ταινίες στις 10 μπορούν για ένα πείραμα τα αποτελέσματα μας αν είναι ίδια και αυτό να μην οφείλεται στο γεγονός ότι ο ταξινομητής μας τα πηγαίνει καλά στις 3 ταινίες αλλά στο γεγονός ότι οι υπόλοιπες 7 είναι τέτοιες ταινίες που δεν μας βοηθούν καθόλου και αν είχαμε επιλέξει διαφορετικές 3 ταινίες να τα πηγαίναμε χάλια. Οπότε προτείνουμε να τρέξετε κάθε φορά έναν αριθμό πειραμάτων για κάθε υποσύνολο ταινιών π.χ. 10 πειράματα με 3 ταινίες, 10 πειράματα για 20 ταινίες κ.ο.κ. και να κρατήσετε σαν τελικό σκορ το μέσο όρο όλων των πειραμάτων. Έτσι τα αποτελέσματα των αγγλικών σας θα είναι πιο αντικειμενικά.

In [ ]:
train_ratings = pd.read_csv(path + "train_ratings.csv")
test_ratings = pd.read_csv(path + "test_ratings.csv")

In [ ]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 5)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5106382978723404
recall: 1.0
f1: 0.676056338028169


In [ ]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 10)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5124670300685773
recall: 0.9888888888888889
f1: 0.6747719477486681


In [ ]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 20)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5137892909063894
recall: 0.9777777777777779
f1: 0.6727605346718513


In [ ]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 30)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5106382978723404
recall: 1.0
f1: 0.676056338028169


In [ ]:
metrics = []
for i in range (10):
    movie_score = train_recommender(train_ratings, rating_weights, score_weights, 40)
    real, pred = predict_example(test_ratings, movie_score)
    metrics.append(get_metrics(real, pred))

for metric in metrics[0].keys():
    print (f"{metric}: {np.mean([m[metric] for m in metrics])}")

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

0it [00:00, ?it/s]

precision: 0.5118680460667577
recall: 0.9791666666666666
f1: 0.6714341296397167


#### Συμπεράσματα

Σχετικά με την ακρίβεια παρατηρούμε ότι με τον πιο μικρό αριθμό ταινιών (5 ταινίες) παίρνουμε την μέγιστη εφικτή ακρίβεια του recommender περίπου στα 51%. 
Το Recall είναι σε υψηλά επίπεδα σε κάθε περίπτωση (98%-100%). Ο recommender μας με βάση τα βάρη που έχουμε ορίσει εμφανίζει τιμή $F_1$ κοντά στα 67%. Προφανώς το precision και συνεπώς και το $F_1$ θα μπορούσαν να αυξηθούν χρησιμοποιώντας πιο σύνθετους κανόνες για τα findSim.
  